src/rust/aggregator/src/pipelines/order_history.rs

```diff
@@ -0,0 +1,324 @@
+        self.process_and_save_internal().await
+    }
+
+    fn poll_interval(&self) -> Option<std::time::Duration> {
+        Some(TIMEOUT)
+    }
+
+    async fn process_and_save_internal(&mut self) -> PipelineAggregationResult {
+        let mut transaction = create_repeatable_read_transaction(&self.pool).await?;
+        let last_indexed_timestamp =
+            sqlx::query_file!("sqlx_queries/order_history/get_last_indexed_timestamp.sql")
+                .fetch_optional(&mut transaction as &mut PgConnection)
+                .await
+                .map_err(to_pipeline_error)?;
+        if last_indexed_timestamp.is_none() {
+            sqlx::query_file!("sqlx_queries/order_history/init_last_indexed_timestamp.sql")
+                .execute(&self.pool)
+                .await
+                .map_err(to_pipeline_error)?;
+        }
+        let last_indexed_timestamp =
+            sqlx::query_file!("sqlx_queries/order_history/get_last_indexed_timestamp.sql")
+                .fetch_optional(&mut transaction as &mut PgConnection)
+                .await
+                .map_err(to_pipeline_error)?;
+        if last_indexed_timestamp.is_none() {
+            return Ok(());
+        }
+        let last_indexed_timestamp = last_indexed_timestamp.unwrap().time;
+        let last_timestamp = sqlx::query_file!("sqlx_queries/order_history/get_last_time_indexed.sql")
+            .fetch_one(&mut transaction as &mut PgConnection)
+            .await
+            .map_err(to_pipeline_error)?
+            .time;
+        let last_timestamp = if let Some(e) = last_timestamp {
+            e
+        } else {
+            return Ok(());
+        };
+        let intervals = (last_timestamp - last_indexed_timestamp).num_minutes() as usize;
+        let mut timestamps = Vec::with_capacity(intervals);
+        for i in 0..intervals {
+            timestamps.push(last_indexed_timestamp + chrono::Duration::minutes(i as i64 + 1));
+        }
+        if timestamps.len() == 0 {
+            return Ok(());
+        }
+        let mut state = self.state.clone().unwrap_or_default();
+        for timestamp in &timestamps {
+            state = if state.orders.is_empty() && state.last_txn_indexed == BigDecimal::zero() {
+                self.get_state_from_timestamp_with_state(
+                    state.orders,
+                    &BigDecimal::zero(),
+                    timestamp,
+                    &mut transaction,
+                )
+                .await?
+            } else {
+                self.get_state_from_timestamp_with_state(
+                    state.orders,
+                    &state.last_txn_indexed,
+                    timestamp,
+                    &mut transaction,
+                )
+                .await?
+            };
+            for order in state.orders.values() {
+                sqlx::query_file!(
+                    "sqlx_queries/order_history/insert_order.sql",
```

Comment:
Could this end up inserting a taker order?

Reply:
Taker orders are closed in the same transaction where they are open right ? If yes, and there's no other logic error in the code, then no.